File: LinkedList/Medium/2487. Remove Nodes From Linked List/2487. Remove Nodes From Linked List.c
#include <stdlib.h>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/**
 * Reverses a singly-linked list.
 * 
 * @param head Pointer to the head of the linked list.
 * @return Pointer to the new head of the reversed list.
 */
struct ListNode* reverse(struct ListNode* head) {
    struct ListNode *prev = NULL, *temp = head, *node;
    while (temp) {
        node = temp->next;
        temp->next = prev;
        prev = temp;
        temp = node;
    }
    return prev;
}

/**
 * Removes nodes from the linked list that have values less than the maximum value
 * of nodes that follow them. The function first reverses the list, then removes
 * nodes based on the criteria, and finally reverses the list again to restore 
 * the original order.
 * 
 * @param head Pointer to the head of the linked list.
 * @return Pointer to the head of the modified list.
 */
struct ListNode* removeNodes(struct ListNode* head) {
    struct ListNode *rev_node;
    head = reverse(head);
    if (head == NULL || head->next == NULL) {
        return head;
    }
    rev_node = head;
    int curr_max_val = head->val;
    while (rev_node->next) {
        if (rev_node->next->val < curr_max_val) {
            rev_node->next = rev_node->next->next;
        } else {
            curr_max_val = rev_node->next->val;
            rev_node = rev_node->next;
        }
    }
    return reverse(head);
}

/**
 * Example usage:
 * struct ListNode* node1 = (struct ListNode*)malloc(sizeof(struct ListNode));
 * node1->val = 1;
 * node1->next = (struct ListNode*)malloc(sizeof(struct ListNode));
 * node1->next->val = 3;
 * node1->next->next = (struct ListNode*)malloc(sizeof(struct ListNode));
 * node1->next->next->val = 2;
 * node1->next->next->next = (struct ListNode*)malloc(sizeof(struct ListNode));
 * node1->next->next->next->val = 4;
 * node1->next->next->next->next = NULL;
 * struct ListNode* new_head = removeNodes(node1); // Expected to remove nodes less than max value
 */

File: LinkedList/Medium/2487. Remove Nodes From Linked List/2487. Remove Nodes From Linked List.py
# Problem: Remove Nodes in a Linked List
# Description: Given a linked list, remove nodes that are less than the maximum value of nodes that follow them.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def removeNodes(self, head):
        """
        Removes nodes from the linked list that have values less than the maximum value
        of nodes that follow them.

        :param head: ListNode, the head of the singly-linked list.
        :return: ListNode, the head of the modified list after removing nodes.
        """
        def reverse(head):
            """
            Reverses the given linked list.

            :param head: ListNode, the head of the singly-linked list.
            :return: ListNode, the head of the reversed list.
            """
            prev = None
            temp = head
            while temp:
                next_node = temp.next
                temp.next = prev
                prev = temp
                temp = next_node
            return prev
        
        # Step 1: Reverse the linked list
        head = reverse(head)
        rev_head = head
        
        # If the list is empty or has only one node, return it as is
        if head is None or head.next is None:
            return head
        
        # Step 2: Traverse the reversed list and remove nodes with values less than the maximum value seen
        curr_max_val = rev_head.val
        while rev_head.next:
            if rev_head.next.val < curr_max_val:
                rev_head.next = rev_head.next.next
            else:
                curr_max_val = rev_head.next.val
                rev_head = rev_head.next
        
        # Step 3: Reverse the list again to restore original order
        return reverse(head)

## Time Complexity: O(N), where N is the number of nodes in the list.
## Space Complexity: O(1), as we do not use extra space for storing nodes.

# Example usage:
# head = ListNode(1)
# head.next = ListNode(2)
# head.next.next = ListNode(3)
# head.next.next.next = ListNode(4)
# solution = Solution()
# new_head = solution.removeNodes(head)  # Expected to remove nodes less than max value
